src/common/src/com/microsoftopentechnologies/tooling/msservices/helpers/auth/AADManagerImpl.java

```diff
@@ -0,0 +1,697 @@
+        authResultLockByUser = new HashMap<UserInfo, ReentrantReadWriteLock>();
+        tempLockByUser = new HashMap<UserInfo, ReentrantReadWriteLock>();
+
+        for (Map.Entry<UserInfo, Map<String, AuthenticationResult>> userAuthResults :
+                authResultByUserResource.entrySet()) {
+            UserInfo userInfo = userAuthResults.getKey();
+            Map<String, AuthenticationResult> authResults = userAuthResults.getValue();
+
+            AuthenticationResult refreshAuthResult = null;
+
+            for (AuthenticationResult authResult : authResults.values()) {
+                if (!StringHelper.isNullOrWhiteSpace(authResult.getRefreshToken()) &&
+                        !StringHelper.isNullOrWhiteSpace(authResult.getResource()) &&
+                        (refreshAuthResult == null || refreshAuthResult.getExpiresOn() < authResult.getExpiresOn())) {
+                    refreshAuthResult = authResult;
+                }
+            }
+
+            if (refreshAuthResult != null) {
+                refreshAuthResultByUser.put(userInfo, refreshAuthResult);
+            }
+
+            authResultLockByUser.put(userInfo, new ReentrantReadWriteLock(false));
+        }
+    }
+
+    @NotNull
+    public static synchronized AADManager getManager() {
+        if (instance == null) {
+            gson = new GsonBuilder().enableComplexMapKeySerialization().create();
+            instance = new AADManagerImpl();
+        }
+
+        return instance;
+    }
+
+    @Override
+    @NotNull
+    public UserInfo authenticate(@NotNull String resource, @NotNull String title)
+            throws AzureCmdException {
+        try {
+            PluginSettings settings = DefaultLoader.getPluginComponent().getSettings();
+            return authenticate(settings.getTenantName(), resource, title, PromptValue.login);
+        } catch (ExecutionException e) {
+            throw new AzureCmdException("Error invoking the authentication process", e.getCause());
+        } catch (Throwable e) {
+            if (e instanceof AzureCmdException) {
+                throw (AzureCmdException) e;
+            }
+
+            throw new AzureCmdException("Error invoking the authentication process", e);
+        }
+    }
+
+    @Override
+    public void authenticate(@NotNull UserInfo userInfo,
+                             @NotNull String resource,
+                             @NotNull String title)
+            throws AzureCmdException {
+        authenticateWithInteractiveToken(userInfo, resource, title);
+    }
+
+    @Override
+    @NotNull
+    public <T> T request(@NotNull UserInfo userInfo,
+                         @NotNull String resource,
+                         @NotNull String title,
+                         @NotNull RequestCallback<T> requestCallback)
+            throws AzureCmdException {
+        if (isAuthenticated(userInfo, resource)) {
+            return authenticatedRequest(userInfo, resource, title, requestCallback);
```

Comment:
Do we have an opportunity here to reduce number of lock accesses and map look-ups by combining the `isAuthenticated` and the `authenticatedRequest` calls somehow using null-checks or Guava's `Optional<T>`? Given that this method is likely to be called with high frequency it might be a worthwhile thing to do.